### PR TITLE
TLV parsing to proxy protocol v2 with AWS VPC extensions.

### PR DESCRIPTION
**Pull request by stu-elastic**

### CODE REVIEW

File: tlv.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	PP2_TYPE_MAX_EXPERIMENT = 0xF7
+	PP2_TYPE_MIN_FUTURE     = 0xF8
+	PP2_TYPE_MAX_FUTURE     = 0xFF
+	// Amazon's extension
+	PP2_TYPE_AWS            = 0xEA
+	PP2_SUBTYPE_AWS_VPCE_ID = 0x01
+)
+
+var (
+	ErrTruncatedTLV    = errors.New("Truncated TLV")
```
Comment (dtuck9) [inline tlv.go:50]: Per Go standards, the first letter of an error string should not be capitalized ([see here](https://github.com/golang/go/wiki/CodeReviewComments#error-strings)).

File: tlv.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+func (p PP2Type) Experiment() bool {
+	return p >= PP2_TYPE_MIN_EXPERIMENT && p <= PP2_TYPE_MAX_EXPERIMENT
+}
+
+// Future is true is the type is reserved for future use, see section 2.2.7
+func (p PP2Type) Future() bool {
+	return p >= PP2_TYPE_MIN_FUTURE
+}
+
+// Spec is true is the type is covered by the spec, see section 2.2 and 2.2.7
```
Comment (dtuck9) [inline tlv.go]: > Spec is true is the type is covered by the spec

Change second "is" to "if"

File: tlv.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+// AWSVPCType is true if the TLV is an AWS extension with VPC subtype
+func (t TLV) AWSVPCType() bool {
+	return t.Type.AWS() && t.Length >= 1 && t.Value[0] == PP2_SUBTYPE_AWS_VPCE_ID
+}
+
+// AWSVPCId returns the vpc-id of an AWS VPC extension TLV or errors with ErrIncompatibleTLV or ErrMalformedTLV if
+// it's the wrong TLV type or has a malformed VPC ID (containing chars other than 0-9, a-z, -)
+func (t TLV) AWSVPCId() (string, error) {
```
Comment (dtuck9) [inline tlv.go]: It pains me to say this, but technically `AWSVPCId` should be `AWSVPCID` per Go standards ([see here](https://github.com/golang/go/wiki/CodeReviewComments#initialisms))

File: tlv.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	for _, c := range t.Value[1:] {
+		if !(c == byte('-') || (c >= byte('0') && c <= byte('9')) || (c >= byte('a') && c <= byte('z'))) {
+			return "", ErrMalformedTLV
+		}
+	}
+	return string(t.Value[1:]), nil
+}
+
+// AWSVPCId returns the first AWS VPC ID in the TLV if it exists and is well-formed.  Returns an empty string otherwise.
+func (header *Header) AWSVPCId() string {
```
Comment (dtuck9) [inline tlv.go]: `AWSVPCId` -> `AWSVPCID`

**Review by dtuck9 — COMMENTED** [review]
Do we have everything we need in place as far as licensing goes since we're forking this from pires/go-proxyproto?

I'm not familiar enough with TLV/proxy protocol/AWS VPC to really review the logic, so I focused on Go idiomatics and comments. Overall looks good.

**Comment by dekimsey** [comment]
I was in the process of investigating the proxyproto library and found this fork off the the pires version. @stu-elastic, I was wondering if you could comment on what is the difference between this pull request and the existing one to add support to TLV [here](https://github.com/pires/go-proxyproto/pull/2)? Do you plan on pushing the changes upstream?

Thanks!

File: tlv.go (hunk 1 above)
Comment (stu-elastic) [reply tlv.go:50]: The existing lib has all errors captialized, so I choose to match the surrounding lib's code.  Eg [	ErrCantReadProtocolVersionAndCommand    = errors.New("Can't read proxy protocol version and command")](https://github.com/elastic/go-proxyproto/pull/1/files#diff-5e68413c67a51bcc7bf6e42ec37bde13L19).  Changing the rest of the code would be an API break, making this less likely to move upstream.

File: tlv.go (hunk 2 above)
Comment (stu-elastic) [reply tlv.go]: Good catch.

File: tlv.go (hunk 3 above)
Comment (stu-elastic) [reply tlv.go]: Noooooo! Yeah I had the same sense as ya did, will change.

File: tlv.go (hunk 4 above)
Comment (stu-elastic) [reply tlv.go]: Changed.

**Comment by stu-elastic** [comment]
@dtuck9 

> Do we have everything we need in place as far as licensing goes since we're forking this from pires/go-proxyproto?

Yup, we're good, staying within the Apache 2.0 license.

> I'm not familiar enough with TLV/proxy protocol/AWS VPC to really review the logic, so I focused on Go idiomatics and comments. Overall looks good.

Here's the relevant part of the [spec](https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt).

```
The information in those bytes will be arranged in Type-Length-Value (TLV
vectors) in the following format.  The first byte is the Type of the vector.
The second two bytes represent the length in bytes of the value (not included
the Type and Length bytes), and following the length field is the number of
bytes specified by the length.

        struct pp2_tlv {
            uint8_t type;
            uint8_t length_hi;
            uint8_t length_lo;
            uint8_t value[0];
        };
```

Essentially we get a type marker and then a big endian uint16 `length` and a series of bytes which are interpreted based on the `type`, eg a 32 bit int for `PP2_TYPE_CRC32C` (`0x03`).

The Amazon extension is [lightly documented](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/load-balancer-target-groups.html#proxy-protocol), they use `PP2_TYPE_AWS`, `0xEA` which is within the range of reserved custom types, `PP2_TYPE_MIN_CUSTOM`, `0xE0` to `PP2_TYPE_MAX_CUSTOM`, `0xEF`.

For `value` they have their own subtypes, presumably to aid in future extensions:
```
struct aws_tlv {
  uint8_t type; // Only PP2_SUBTYPE_AWS_VPCE_ID 0x01 currently
  uint8_t value[0]; // Interpreted as string for PP2_SUBTYPE_AWS_VPCE_ID
}
```

This means that their length in the outer struct, `pp2_tlv` with type `0xEA`, is 1 + length of the the VPC ID.  The extra byte is for their subtype.

I've copied over their [unit test](https://github.com/aws/elastic-load-balancing-tools/blob/c8eee30ab991ab4c57dc37d1c58f09f67bd534aa/proprot/tst/com/amazonaws/proprot/Compatibility_AwsNetworkLoadBalancerTest.java#L41..L67), also Apache 2.0, to ensure we're parsing it the same as in their sample Java lib.

**Comment by stu-elastic** [comment]
@dekimsey in addition to TLV parsing, that change was a major refactoring of the package, it includes a connection wrapper that hides the proxy protocol header values under `RemoteAddr`, resulting in a significant API break.

This change simply extends the existing API to add `TLVs` to the `Header struct`, adding a field which is only populated if TLVs exist in the V2 version of the protocol.  It is not an API break, users of pires's version who don't care about TLVs can simply ignore the change.

I intend to submit this change upstream and will delete our fork if it, or equivalent functionality, is merged.

**Comment by Freeaqingme** [comment]
I'm the author of the PR that was referred to. As has been mentioned under the discussion of that PR it's gone a bit stale, so I definitely support any improvement wrt TLV support.

Was this also tested with other TLV implementations (e.g. haproxy)?

File: tlv.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	rest, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+	var tlvs []TLV
+	for i := 0; i < len(rest); {
+		tlv := TLV{
+			Type: PP2Type(rest[i]),
+		}
+		tlv.Length = int(binary.BigEndian.Uint16(rest[i+1 : i+3])) // Max length = 65K
```
Comment (dtuck9) [inline tlv.go]: What happens if `i+1` or `i+3` is outside of index range of `rest`? Would probably be worthwhile to have some unit tests around edge cases.

File: tlv.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+// AWSVPCID returns the vpc-id of an AWS VPC extension TLV or errors with ErrIncompatibleTLV or ErrMalformedTLV if
+// it's the wrong TLV type or has a malformed VPC ID (containing chars other than 0-9, a-z, -)
+func (t TLV) AWSVPCID() (string, error) {
+	if !t.AWSVPCType() {
+		return "", ErrIncompatibleTLV
+	}
+	for _, c := range t.Value[1:] {
+		if !(c == byte('-') || (c >= byte('0') && c <= byte('9')) || (c >= byte('a') && c <= byte('z'))) {
```
Comment (dtuck9) [inline tlv.go]: Are upper case chars not valid? I haven't been able to find the documentation for it.

Would it make sense to have regex for this?

**Review by dtuck9 — COMMENTED** [review]
2 questions regarding implementation

**Comment by dekimsey** [comment]
> I intend to submit this change upstream and will delete our fork if it, or equivalent functionality, is merged.

Awesome @stu-elastic! Thank you.

File: tlv.go (hunk 5 above)
Comment (stu-elastic) [reply tlv.go]: Great catch, added check, will add test.

File: tlv.go (hunk 6 above)
Comment (stu-elastic) [reply tlv.go]: Yeah there's no docs for the format of amazon resource ids, the defacto standard is `a-z0-9-`, I've changed to a regex that allows caps as well.

**Comment by stu-elastic** [comment]
@Freeaqingme 
> Was this also tested with other TLV implementations (e.g. haproxy)?

I've tested it with AWS NLBs and HAProxy version 1.9.5

**Comment by stu-elastic** [comment]
@dtuck9 

I've changed the API around a bit.  Previously, invalid TLVs would return an error, even if the consumer didn't care about TLVs.  This would be an API break since previously working use cases would start breaking.

Now, we store the raw TLV bytes and, if the consumer desires, they can try to split and interpret the TLVs, handling any errors that may occur.

**Comment by dtuck9** [comment]
LGTM

File: tlv.go (hunk 5 above)
Comment (stu-elastic) [reply tlv.go]: Added.  I actually couldn't get this to fail because the underlying capacity was so large (from `ioutil.ReadAll`) that it wasn't OOBing.

File: header.go [hunk 7]
```diff
@@ -72,6 +72,9 @@ func (header *Header) EqualsTo(otherHeader *Header) bool {
 	if header.Command.IsLocal() {
 		return true
 	}
+	if header.Version == 0x02 && !bytes.Equal(header.rawTLVs, otherHeader.rawTLVs) {
```
Comment (dtuck9) [inline header.go:76]: Why does it matter what version the header is? Perhaps add a comment?

File: tlv.go [hunk 8]
```diff
@@ -51,7 +51,7 @@ var (
 	ErrMalformedTLV    = errors.New("Malformed TLV Value")
 	ErrIncompatibleTLV = errors.New("Incompatible TLV type")
 
-	vpceRe = regexp.MustCompile("[A-Za-z0-9-]*")
+	vpceRe = regexp.MustCompile("^[A-Za-z0-9-]*$")
```
Comment (dtuck9) [inline tlv.go:54]: Good catch

**Review by dtuck9 — APPROVED** [review]
Still LGTM

File: header.go (hunk 7 above)
Comment (stu-elastic) [reply header.go:76]: TLVs only exist for version 2, will do.

File: tlv.go (hunk 8 above)
Comment (stu-elastic) [reply tlv.go:54]: Hurrah for unit tests.

File: header.go (hunk 7 above)
Comment (stu-elastic) [reply header.go:76]: Added.

**Comment by stu-elastic** [comment]
https://github.com/pires/go-proxyproto/pull/9
